manifests/init.pp

```diff
@@ -34,21 +34,39 @@
     Hash $pkg_list,
 ) {
 
-    # Default value (for bare package names)
-    $default = {'ensure' => 'installed'}
+  $script_name="puppet-profile_additional_packages-packages_without_params.sh"
 
-    # Limit list of packages by OS Family
-    $packages = $pkg_list[ $facts['os']['family'] ]
+  file { "/root/scripts/${script_name}":
+    ensure  => 'present',
+    mode    => '0740',
+    owner   => 'root',
+    group   => 'root',
+    source  => "puppet:///modules/${module_name}/root/scripts/${script_name}",
+    require => File['/root/scripts'],
+  }
 
-    if $packages =~ Hash[ String[1], Data, 1 ] {
-      # Find keys without a value and add default value
-      $sane_pkg_list = $packages.map |$key, $val| {
-          if $val { [ $key, $val ] }
-          else { [ $key, $default ] }
-      }.convert_to(Hash)
+  # Limit list of packages by OS Family
+  $packages = $pkg_list[ $facts['os']['family'] ]
 
-      ensure_packages( $sane_pkg_list )
+  if $packages =~ Hash[ String[1], Data, 1 ] {
+
+    # Find keys without a value and install those pacakges
+    # via exec/'yum install' (single yum transaction)
+    $packages_without_params = $packages.filter |$key, $val| {
+      empty($val)
+    }.keys.join(' ')
+    exec { 'install_packages_without_params':
+      command => "/root/scripts/${script_name} ${packages_without_params}",
+      unless  => "/usr/bin/rpm -q ${packages_without_params}" 
     }
```

Comment:
I can definitely change this to yum, no problem.

The script determines which of the packages are not installed and:
(1) passes only those to dnf(-> yum), so that 'yum history' will reflect exactly which packages are getting installed; e.g., if the node's already been up for a while and we add one package, it will only try to install that and the yum history for that transaction will only show that package rather than the whole list;
(2) uses logger to log only those that are being installed (not the whole list of packages w/o params) to syslog

Reply:
That makes sense about the script doing a bit more. I like that.

Reply:
Updated to use yum instead.